agent/proxycfg/state.go

```diff
@@ -0,0 +1,357 @@
+			if err != nil {
+				s.logger.Printf("[ERR] Failed to copy config snapshot for proxy %s",
+					s.proxyID)
+				continue
+			}
+			replyCh <- snapCopy
+
+			// Skip rest of loop - there is nothing to send since nothing changed on
+			// this iteration
+			continue
+		}
+
+		// Check if snap is complete enough to be a valid config to deliver to a
+		// proxy yet.
+		if snap.Valid() {
+			// Don't send it right away, set a short timer that will wait for updates
+			// from any of the other cache values and deliver them all together.
+			if coallesceTimer == nil {
+				coallesceTimer = time.AfterFunc(coallesceTimeout, func() {
+					// This runs in another goroutine so we can't just do the send
+					// directly here as access to snap is racy so just signal the main
+					// loop above.
+					sendCh <- struct{}{}
+				})
+			}
+		}
+	}
+}
+
+func (s *State) handleUpdate(u cache.UpdateEvent, snap *ConfigSnapshot) error {
+	switch u.CorrelationID {
+	case "roots":
+		roots, ok := u.Result.(*structs.IndexedCARoots)
+		if !ok {
+			return fmt.Errorf("invalid type for roots response: %T", u.Result)
+		}
+		snap.Roots = roots
+	case "leaf":
+		leaf, ok := u.Result.(*structs.IssuedCert)
+		if !ok {
+			return fmt.Errorf("invalid type for leaf response: %T", u.Result)
+		}
+		snap.Leaf = leaf
+	case "intentions":
+		// Not in snapshot currently, no op
+	default:
+		// Service discovery result, figure out which type
+		if strings.HasPrefix(u.CorrelationID, "service:") {
+			resp, ok := u.Result.(*structs.IndexedCheckServiceNodes)
+			if !ok {
+				return fmt.Errorf("invalid type for service response: %T", u.Result)
+			}
+			snap.UpstreamEndpoints[u.CorrelationID] = resp.Nodes
+		} else if strings.HasPrefix(u.CorrelationID, "prepared_query:") {
+			resp, ok := u.Result.(*structs.PreparedQueryExecuteResponse)
+			if !ok {
+				return fmt.Errorf("invalid type for prepared query response: %T", u.Result)
+			}
+			snap.UpstreamEndpoints[u.CorrelationID] = resp.Nodes
+		} else {
+			return errors.New("unknown correlation ID")
+		}
+	}
+	return nil
+}
+
+// CurrentSnapshot synchronously returns the current ConfigSnapshot if there is
+// one ready. If we don't have one yet because not all necessary parts have been
+// returned (i.e. both roots and leaf cert), nil is returned.
+func (s *State) CurrentSnapshot() *ConfigSnapshot {
+	// Make a chan for the response to be sent on
+	ch := make(chan *ConfigSnapshot, 1)
+	s.reqCh <- ch
+	// Wait for the response
+	return <-ch
+}
+
+// Close discards the state and stops any long-running watches.
+func (s *State) Close() error {
+	if s.cancel != nil {
+		s.cancel()
+		s.cancel = nil
+	}
+	if s.snapCh != nil {
+		close(s.snapCh)
+		s.snapCh = nil
+	}
+	return nil
+}
+
+// Changed returns whether or not the passed NodeService has had any of the
+// fields we care about for config state watching changed or a different token.
+func (s *State) Changed(ns *structs.NodeService, token string) bool {
```

Comment:
Maybe I missed it, but I would add a table test for this. This seems like something that we should test the edges for and be prepared for change in the future as well.

Reply:
Good catch.